### PR TITLE
[FEATURE]: created the bff part of the authentication

### PR DESCRIPTION
**Pull request by maelaubert56**
**Branch:** feature/ISSUE-7/authentication

**Commits:**
- [FEATURE] Added the Account Provider and Model to 'commons']
- [FEATURE]: Account queries and mutations in BFF
- [FEATURE]: Account mutations in BFF
- [FEATURE]: Added password & jwt utils and Login Query
- [FEATURE]: Token authentication

**Pull Request title:** [FEATURE]: created the bff part of the authentication

**Description:**
This PR contains the queries and mutations for the authentication. It also secure the bff for non authenticated users.

### CODE REVIEW

File: src/v6y-bff/src/index.ts [hunk 1]
```diff
@@ -59,9 +67,20 @@ app.use(
...
+                req.body?.operationName !== 'IntrospectionQuery'
+            ) {
+                const user = await passportAuthenticate(req);
+                if (!user) {
+                    throw new Error('Unauthorized');
+                }
+                return { user };
+            }
+            return {};
+        },
```
Comment (maelaubert56) [inline src/v6y-bff/src/index.ts:83]: create an authentication middleware

File: src/v6y-bff/src/resolvers/account/AccountMutations.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        const whereClause = params?.input?.where;
+        if (!whereClause) {
+            return null;
+        }
+
+        const accountId = whereClause._id;
+        AppLogger.info(`[AccountMutations - deleteAccount] accountId : ${accountId}`);
+
+        await AccountProvider.deleteAccount({ _id: accountId });
+
```
Comment (maelaubert56) [inline src/v6y-bff/src/resolvers/account/AccountMutations.ts:93]: log the deleted status

File: src/v6y-bff/src/resolvers/account/AccountQueries.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        if (!accountDetails || !accountDetails.password || !accountDetails._id) {
+            return null;
+        }
+
+        const isPasswordMatch = await validatePassword(password, accountDetails.password);
+
+        if (!isPasswordMatch) {
+            return null;
+        }
+        const token = passportGenerateToken(accountDetails);
```
Comment (maelaubert56) [inline src/v6y-bff/src/resolvers/account/AccountQueries.ts:105]: add a line

File: src/v6y-commons/package.json [hunk 4]
```diff
@@ -26,6 +26,9 @@
   "devDependencies": {
     "@eslint/js": "=9.11.1",
     "@types/eslint__js": "=8.42.3",
+    "@types/jsonwebtoken": "^9.0.7",
+    "@types/passport": "^1.0.16",
+    "@types/passport-jwt": "^4.0.1",
```
Comment (maelaubert56) [inline src/v6y-commons/package.json:31]: fix versions

File: src/v6y-commons/package.json [hunk 5]
```diff
@@ -47,13 +50,17 @@
     "adm-zip": "=0.5.10",
     "axios": "=1.6.2",
     "axios-retry": "=4.0.0",
+    "bcrypt": "^5.1.1",
```
Comment (maelaubert56) [inline src/v6y-commons/package.json:53]: fix version

File: src/v6y-commons/src/core/PassportUtils.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
+import jwt from 'jsonwebtoken';
+import passport from 'passport';
+import { ExtractJwt, Strategy as JwtStrategy } from 'passport-jwt';
+
+import AccountProvider from '../database/AccountProvider.ts';
+import { AccountType } from '../types/AccountType.ts';
+import AppLogger from './AppLogger.ts';
+
+const SECRET_KEY = process.env.JWT_SECRET || 'temporaryToken'; // TODO: Remove this static tokens
```
Comment (maelaubert56) [inline src/v6y-commons/src/core/PassportUtils.ts:9]: remove 'temporaryToken'

File: src/v6y-commons/src/core/PassportUtils.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+if (SECRET_KEY === undefined) {
+    throw new Error('JWT_SECRET is not defined in the environment variables');
+}
+
+const opts = {
```
Comment (maelaubert56) [inline src/v6y-commons/src/core/PassportUtils.ts:15]: jwtOptions instead of opts

File: src/v6y-commons/src/core/PassportUtils.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+    secretOrKey: SECRET_KEY,
+};
```
Comment (maelaubert56) [inline src/v6y-commons/src/core/PassportUtils.ts:18]: move into passport.use()

File: src/v6y-commons/src/core/PassportUtils.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+export const passportAuthenticate = (request: Request) =>
+    new Promise((resolve) => {
+        passport.authenticate('jwt', { session: false }, (err: Error | null, user: unknown) => {
+            if (err || !user) {
+                AppLogger.info(`[ApolloServer] Not authenticated : ${err || 'No user found'}`);
+                resolve(null);
+            } else {
+                resolve(user);
+            }
+        })(request);
```
Comment (maelaubert56) [inline src/v6y-commons/src/core/PassportUtils.ts:53]: async await instead of a Promise

File: src/v6y-commons/src/core/PassportUtils.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    new JwtStrategy(opts, async (jwtPayload, done) => {
+        try {
+            // Vérifier si le token contient bien un _id ou email
+            if (!jwtPayload._id) {
+                return done(null, false);
+            }
+
+            const accountDetails = await AccountProvider.getAccountDetailsByParams({
+                _id: jwtPayload._id,
+            });
```
Comment (maelaubert56) [inline src/v6y-commons/src/core/PassportUtils.ts:30]: put this in a function getUser(_id) in a file 'AuthenticationManager'

File: src/v6y-commons/src/database/AccountProvider.ts [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+/**
+ * Create An Account
+ * @param account
+ */
+const createAccount = async (account: AccountInputType) => {
+    try {
+        const createdAccount = (await AccountModelType.create(account))?.dataValues;
+        AppLogger.info(`[AccountProvider - createAccount] createdAccount: ${createdAccount._id}`);
+        return createdAccount;
```
Comment (maelaubert56) [inline src/v6y-commons/src/database/AccountProvider.ts:54]: add line

File: src/v6y-commons/src/database/AccountProvider.ts [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    } catch (error) {
+        AppLogger.info(`[AccountProvider - createAccount] error:  ${error}`);
+        return null;
```
Comment (maelaubert56) [inline src/v6y-commons/src/database/AccountProvider.ts:57]: add line

File: src/v6y-commons/src/database/AccountProvider.ts [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+    }
+};
+
+/**
+ * Delete an Account
+ * @param _id
+ */
+const deleteAccount = async ({ _id }: AccountType) => {
+    try {
```
Comment (maelaubert56) [inline src/v6y-commons/src/database/AccountProvider.ts:93]: log account id

File: src/v6y-commons/src/database/AccountProvider.ts [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+
+    return queryOptions;
+};
+
+/**
+ * Create An Account
+ * @param account
+ */
+const createAccount = async (account: AccountInputType) => {
+    try {
```
Comment (maelaubert56) [inline src/v6y-commons/src/database/AccountProvider.ts:51]: log params

File: src/v6y-commons/src/database/AccountProvider.ts [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+        if (!_id) {
+            return null;
+        }
+
+        await AccountModelType.destroy({
+            where: {
+                _id,
+            },
+        });
+
```
Comment (maelaubert56) [inline src/v6y-commons/src/database/AccountProvider.ts:103]: log deleted succes

File: src/v6y-commons/src/database/AccountProvider.ts [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+            `[AccountProvider - getAccountDetailsByParams] _id: ${_id}, email: ${email}`,
+        );
+        if (!_id && !email) {
+            return null;
+        }
+
+        const accountDetails = await AccountModelType.findOne({
+            where: _id ? { _id } : { email },
+        });
+
```
Comment (maelaubert56) [inline src/v6y-commons/src/database/AccountProvider.ts:130]: log account id found

File: src/v6y-commons/src/database/AccountProvider.ts [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+        AppLogger.info(
+            `[AccountProvider - getAccountListByPageAndParams] searchQuery: ${JSON.stringify(searchQuery)}`,
+        );
+
+        const accounts = await AccountModelType.findAll();
+
+        AppLogger.info(
+            `[AccountProvider - getAccountListByPageAndParams] accountList: ${accounts?.length}`,
+        );
+        return accounts?.map((item) => item?.dataValues) || [];
```
Comment (maelaubert56) [inline src/v6y-commons/src/database/AccountProvider.ts:180]: add line

**Review by helabenkhalfallah — APPROVED** [review]
Thank you so much! 👍  🥇

**Pull request merged**